runners: jlink: Add support for thread info

carlescufi opened the pull request:
Starting with J-Link v7.11b, The J-Link Software and Documentation Pack
supports Zephyr RTOS awareness, aka thread info.
For this to work, the Zephyr application needs to be built with
CONFIG_DEBUG_THREAD_INFO=y.

This patch adds the corresponding JLinkGDBServer option when invoking
it, in order to activate the plugin.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

MaureenHelm commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -153,20 +153,29 @@ def supports_nogui(self):
 
+    def supports_thread_info(self):
+        # The RTOSPlugin_Zephyr was introduced in 7.11b
+        return version.parse(ver) >= version.parse("7.11")
```
You need to read the version before you parse it.
```
$ west -vv debug
ZEPHYR_BASE=/home/maureen/zephyrproject/zephyr (origin: configfile)
-- west debug: rebuilding
cmake version 3.13.2 is OK; minimum version is 3.13.1
Running CMake: /home/maureen/.local/bin/cmake --build /home/maureen/zephyrproject/zephyr/build
ninja: no work to do.
-- west debug: using runner jlink
runner_config: RunnerConfig(build_dir='/home/maureen/zephyrproject/zephyr/build', board_dir='/home/maureen/zephyrproject/zephyr/boards/arm/mimxrt685_evk', elf_file='/home/maureen/zephyrproject/zephyr/build/zephyr/zephyr.elf', hex_file=None, bin_file='/home/maureen/zephyrproject/zephyr/build/zephyr/zephyr.bin', gdb='/home/maureen/.local/zephyr-sdk-0.12.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gdb', openocd='/home/maureen/.local/zephyr-sdk-0.12.2/sysroots/x86_64-pokysdk-linux/usr/bin/openocd', openocd_search='/home/maureen/.local/zephyr-sdk-0.12.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts')
Traceback (most recent call last):
  File "/home/maureen/.local/bin/west", line 11, in <module>
    sys.exit(main())
  File "/home/maureen/.local/lib/python3.6/site-packages/west/app/main.py", line 779, in main
    app.run(argv or sys.argv[1:])
  File "/home/maureen/.local/lib/python3.6/site-packages/west/app/main.py", line 106, in run
    self.run_command(argv)
  File "/home/maureen/.local/lib/python3.6/site-packages/west/app/main.py", line 338, in run_command
    self.run_extension(args.command, argv)
  File "/home/maureen/.local/lib/python3.6/site-packages/west/app/main.py", line 408, in run_extension
    command.run(args, unknown, self.topdir, manifest=self.manifest)
  File "/home/maureen/.local/lib/python3.6/site-packages/west/commands.py", line 129, in run
    self.do_run(args, unknown)
  File "/home/maureen/zephyrproject/zephyr/scripts/west_commands/debug.py", line 34, in do_run
    do_run_common(self, my_args, runner_args)
  File "/home/maureen/zephyrproject/zephyr/scripts/west_commands/run_common.py", line 228, in do_run_common
    runner.run(command_name)
  File "/home/maureen/zephyrproject/zephyr/scripts/west_commands/runners/core.py", line 491, in run
    self.do_run(command, **kwargs)
  File "/home/maureen/zephyrproject/zephyr/scripts/west_commands/runners/jlink.py", line 166, in do_run
    if self.supports_thread_info():
  File "/home/maureen/zephyrproject/zephyr/scripts/west_commands/runners/jlink.py", line 158, in supports_thread_info
    return version.parse(ver) >= version.parse("7.11")
NameError: name 'ver' is not defined
```

carlescufi replied:
Thanks, pushed an old version, now fixed.

carlescufi replied:
Also caching it now, to avoid checking multiple times.

carlescufi commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -121,6 +122,7 @@ def print_gdbserver_message(self):
         self.logger.info('J-Link GDB server running on port {}'.
                          format(self.gdb_port))
 
+    @lru_cache(maxsize=1)
```
It is probably more optimal to cache using a variable instead, but compared to the overhead of executing a process I felt this wasn't worth it.

gmarull replied:
is it worth doing that here? do we call this funcion frequently? version could also be cached the first time it is read or on init

carlescufi replied:
As it is, it spawns a new process and reads its output, so it's very expensive. We need to cache its result, which is immutable, one way or another, and this is the smallest way to do it in terms of lines of code.

carlescufi replied:
>version could also be cached the first time it is read
That is exactly what this line is doing

gmarull commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -91,7 +91,7 @@ def is_extended_samba_protocol(self):
         build_conf = BuildConfiguration(self.cfg.build_dir)
 
-        if 'CONFIG_USE_DT_CODE_PARTITION' not in build_conf:
+        if build_conf.get('CONFIG_USE_DT_CODE_PARTITION', 'n') != 'y':
```
This should also work:
```suggestion
        if build_conf.get('CONFIG_USE_DT_CODE_PARTITION') != 'y':
```

gmarull commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -127,7 +127,7 @@ def get_board_name(self):
         build_conf = BuildConfiguration(self.cfg.build_dir)
 
-        if build_conf['CONFIG_HAS_FLASH_LOAD_OFFSET']:
+        if build_conf.get('CONFIG_HAS_FLASH_LOAD_OFFSET', 'n') == 'y':
```
```suggestion
        if build_conf.get('CONFIG_HAS_FLASH_LOAD_OFFSET') == 'y':
```

gmarull commented on scripts/west_commands/runners/core.py:
```diff
@@ -472,14 +472,23 @@ def get_flash_address(cls, args: argparse.Namespace,
         returned.'''
         if args.dt_flash:
-            if build_conf['CONFIG_HAS_FLASH_LOAD_OFFSET']:
+            if build_conf.get('CONFIG_HAS_FLASH_LOAD_OFFSET', 'n') == 'y':
```
```suggestion
            if build_conf.get('CONFIG_HAS_FLASH_LOAD_OFFSET') == 'y':
```

gmarull commented on scripts/west_commands/runners/core.py:
```diff
@@ -472,14 +472,23 @@ def get_flash_address(cls, args: argparse.Namespace,
+    def thread_info_enabled(cls, build_conf: BuildConfiguration) -> bool:
+        ''' Returns whether debug thread info is enabled in the build. '''
+
+        dti = build_conf.get('CONFIG_DEBUG_THREAD_INFO', 'n')
+        oos = build_conf.get('CONFIG_OPENOCD_SUPPORT', 'n')
```
```suggestion
        dti = build_conf.get('CONFIG_DEBUG_THREAD_INFO')
        oos = build_conf.get('CONFIG_OPENOCD_SUPPORT')
```

gmarull commented on scripts/west_commands/sign.py:
```diff
@@ -154,7 +154,7 @@ def do_run(self, args, ignored):
         # Decide on output formats.
         formats = []
-        bin_exists = 'CONFIG_BUILD_OUTPUT_BIN' in bcfg
+        bin_exists = bcfg.get('CONFIG_BUILD_OUTPUT_BIN', 'n') == 'y'
```
```suggestion
        bin_exists = bcfg.get('CONFIG_BUILD_OUTPUT_BIN') == 'y'
```

gmarull commented on scripts/west_commands/sign.py:
```diff
@@ -164,7 +164,7 @@ def do_run(self, args, ignored):
             formats.append('bin')
 
-        hex_exists = 'CONFIG_BUILD_OUTPUT_HEX' in bcfg
+        hex_exists = bcfg.get('CONFIG_BUILD_OUTPUT_HEX', 'n') == 'y'
```
```suggestion
        hex_exists = bcfg.get('CONFIG_BUILD_OUTPUT_HEX') == 'y'
```

mbolivar-nordic committed:
runners: tests: fix os.path.isfile patch

Copy a fix from test_nrfjprog.py to the other runner test suites. The
current code will enter an infinite recursion if you hit the path
where os.path.isfile is called, since it's been patched to
os_path_isfile_patch in the calling context. The fix is to cache the
'real' version in the parent scope and call it directly as a fallback.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: replace version parsing code

JLink versions like 'V7.0a' do not conform to PEP 440 version
conventions; the 'a' part is used by PEP 440 compliant versions for
alphas. It gets parsed to a legacy type by the packaging library,
which always is treated as a lower value when compared with a
conforming version string.

To fix, get the version from the shared library distributed with the
JLink tools. This has the side benefit of making the code work on
Windows. That's merely a nice to have for -nogui 1 detection for now,
but will be essential in the next commit.

Reported-by: Jake Mercer <jake.mercer@civica.co.uk>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: return shutil.which() value from require()

Make it easy to get the full path to a required binary.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
requirements-base.txt: add pylink

This will be used to deal with the Segger shared library in a portable
way in the jlink runner.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: fix --commander help

The default commander path is platform specific.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: try to improve --dt-flash option help

This option has existed since the beginning of the runners package,
which greatly predates the way DT is used in zephyr right now. It
never really worked the way I wanted it to but it's too much work to
fix it now. Try to improve the help a bit at least while I'm looking
at it again.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: clean up commander script logging

Put it all in one log message rather than splitting it up.

This makes it look cleaner now that each log message is prefixed with
'runners.link:'.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: refactor helper for getting flash address

Separate the logic that gets the right address from .config
into its own helper.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: rework --dt-flash handling

Defer loading .config until we really need it, when we are flashing a
binary. Pre-emptively loading it is wasted effort if we're flashing a
.hex, which has been the default behavior when possible since
dcaabb860f ("west: runners: jlink: prefer .hex over .bin").

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: add **kwargs to other subprocess wrappers

Just to make these match check_output().

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: silence commander output by default

This is really verbose, and I doubt anyone cares unless there is a
problem. Keep it around when run as 'west -v flash' to allow for
debugging, though.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: re-work BuildConfiguration boolean handling

Set options that are definitely true or false to True or False in the
options dict. Add a 'getboolean' method that also allows a fallback to
False in case the option is not mentioned in .config due to unmet
dependencies. This allows calling code to just ask about the option
they are interested in, even if the .config file doesn't mention the
option at all.

Propagate this to users within the runners package and 'west sign',
taking advantage of the new build_conf property.

Rename the 'bcfg' internal variable in sign.py to 'build_conf' to
match other source files that use BuildConfiguration instances, to
make it easier to grep for users.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: add build_conf property

Make it easier to get a BuildConfiguration from runner code.
Stash the result so it only has to be computed once.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: tests: don't patch BuildConfiguration

Instead of mocking out the BuildConfiguration class, just create its
input file and let the real class do the work.

This in turn exposes a bug in the way the board name is being pulled
out of the BuildConfiguration, which we fix to keep the tests passing.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: core: add thread_info_enabled property

This makes it easier for runners to check if the binary has thread
info support turned on, allowing automatic configuration of the
underlying tool to support threads, if possible.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: extend 'nogui' support

Support this when debugging also.

Tweak the style for brevity also while we're here by propertizing the
supports_nogui method, etc.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
runners: jlink: support thread_info_enabled

Automatically enable per-thread info in GDB if the binary and J-Link
support it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@carlescufi I've pushed a version here which should fix all the BuildConfiguration issues and includes this support as well:

https://github.com/mbolivar-nordic/zephyr/tree/jlink-rtos

It also fixes J-Link version detection on Windows, which we will need for this PR.

However, I can't get thread info printed reliably, even though I can see the commands I want are being run, like so: https://gist.github.com/mbolivar-nordic/0add8fa4e10deabbfc21ecca12c3ca5a

It does work sometimes, though, and I get something like this at the end: https://gist.github.com/mbolivar-nordic/8e62ba863f918c2d2b89532143868843

Could this be a bug in the plugin?

mbolivar-nordic commented:
> Could this be a bug in the plugin?

Thread detection seems to work much more reliably when I test on my Windows machine than on my Linux box.

mbolivar-nordic commented:
Thanks to @carlescufi for helping me figure out what was wrong on my Linux box. This is working for me now.

mbolivar-nordic commented:
@MaureenHelm @gmarull please revisit your reviews; I've rewritten this since the original version.

gmarull commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -268,16 +270,13 @@ def flash(self, **kwargs):
-
-            cmd = ([self.commander] + nogui +
+            cmd = ([self.commander] +
+                   (['-nogui', '1'] if self.supports_nogui else []) +
                    ['-if', self.iface,
                     '-speed', self.speed,
                     '-device', self.device,
                     '-CommanderScript', fname] +
+                   (['-nogui', '1'] if self.supports_nogui else []) +
```
repeated `-nogui` option

mbolivar-nordic replied:
> repeated `-nogui` option

fixed, thanks

nandojve commented:
Hello @mbolivar-nordic ,
Nice improvement. Let me run some tests to make sure everything is OK.

henrikbrixandersen commented:
@mbolivar, @carlescufi: Thank you for providing this!
